mlx/primitives.cpp

```diff
@@ -813,20 +813,62 @@ std::vector<array> Convolution::vjp(
         padding_hi[i] = in_size - out_size + padding_[i];
       }
 
+      // Check for negative padding
+      bool has_neg_padding = false;
+      for (auto& pd : padding_lo) {
+        has_neg_padding = (pd < 0) || has_neg_padding;
+      }
+      for (auto& pd : padding_hi) {
+        has_neg_padding = (pd < 0) || has_neg_padding;
+      }
+
+      auto padding_lo_ = std::vector<int>(padding_lo);
+      auto padding_hi_ = std::vector<int>(padding_hi);
+
+      // Use negative padding on the gradient output
+      if (has_neg_padding) {
+        for (auto& p : padding_lo_) {
+          p = std::max(0, p);
+        }
+        for (auto& p : padding_hi_) {
+          p = std::max(0, p);
+        }
+      }
```

Comment:
Yes, this is needed. In some combinations of kernel_size/padding, applying the negative padding before performing the convolution (as done in `conv_general`) leads to wrong gradients w.r.t. the input. E.g., on current master, this will fail without above changes (you need the `|=` fix or it will fail for another reason):
```
import mlx.core as mx
import torch
import numpy as np

x = mx.random.normal((1,8,8,1))
w = mx.random.normal((1,1,1,1))
ct = mx.random.normal((1,5,5,1))

_, grad_mx = mx.vjp(lambda x, w: mx.conv2d(x, w, padding=1, stride=2), [x, w], [ct])
mx_grad_in = grad_mx[0]
pt_grad_in = torch.nn.functional.grad.conv2d_input(
    (1,1,8,8),
    torch.from_numpy(np.array(w)).permute(0,-1,1,2),
    torch.from_numpy(np.array(ct)).permute(0,-1,1,2),
    padding=1,
    stride=2).permute(0, 2, 3, 1)

for i in range(4):
    assert mx_grad_in[i] == pt_grad_in[i], "input grad shape mismatch"
```
One could argue that `kernel_size = 1` and `stride = 2` is unusual, but with my changes, it matches pytorch.

Reply:
Right, you did add the `|=` fix to `conv_general` though
So what is this block doing now that isn't happening later in the `conv_general` with the `|=` fix ?

Reply:
The code only passes positive padding to `conv_general` and applies any negative padding afterwards. Without that, the gradient w.r.t. the input will have a wrong shape in some kernel size/stride/padding combinations (as shown in my example).
You can try out to only apply the `|=` fix and then run my example. You’ll see that the input gradient won’t have the same shape as the input, which is clearly incorrect.

Reply:
Ooh, I see it now 
The issue is with the way we are doing negative padding when the input is dilated 
I'll write up a general fix for that instead of trying to step around that error

Reply:
We can merge this for now though

Reply:
Ok, fair enough. I fixed the failing test. I think it's ready to merge.